### PR TITLE
Issue #56: fix NPathComplexityCheck

### PR DESCRIPTION
**Pull request by kazachka**
Issue #56

Continuation of Pull #2914 .

### CODE REVIEW

**Comment by rnveach** [comment]
@kazachka 
> Continuation of Pull #2914 .

This completely replaces the other PR so we can close it?
Does this one fix the 14 items @romani made in the last PR? I still see some comments that say `Spatial`.

**Comment by kazachka** [comment]
First 4 of 14 were done earlier. I fix other items with except of 10:
> please explain in comment what is "2".
I don't know what means that 2.

I change only SpatialRange class name and didn't change comments.

File: config/suppressions.xml [hunk 1]
```diff
@@ -29,6 +29,7 @@
 
     <!-- Tone down the checking for test code -->
     <suppress checks="CyclomaticComplexity" files="[\\/]XDocsPagesTest\.java"/>
+    <suppress checks="NPathComplexity" files="[\\/]XDocsPagesTest\.java"/>
```
Comment (romani) [inline config/suppressions.xml]: I do not see update in this file.
Why you need this ?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java [hunk 2]
```diff
@@ -102,21 +106,24 @@ public void setMax(int max) {
...
 
     @Override
     public int[] getRequiredTokens() {
-        return new int[] {
-            TokenTypes.CTOR_DEF,
-            TokenTypes.METHOD_DEF,
-            TokenTypes.INSTANCE_INIT,
-            TokenTypes.STATIC_INIT,
-        };
+        return getAcceptableTokens();
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java]: this looks like a separate issue, that not all tokens are in use, that is error prone, lets make separate issue on this and merge separately and quicker.

File: config/suppressions.xml (hunk 1 above)
Comment (rnveach) [inline config/suppressions.xml]: @romani see discussion in https://github.com/checkstyle/checkstyle/issues/2906

Comment (romani) [inline config/suppressions.xml]: Right, ok

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java (hunk 2 above)
Comment (kazachka) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java]: that was in https://github.com/checkstyle/checkstyle/pull/2914#issuecomment-185746384

> I think we should call getAcceptableTokens from getRequiredTokens.

Comment (kazachka) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java]: getDefaultTokens reverted and xdocs changed

**Comment by kazachka** [comment]
javadocs reverted

**Comment by romani** [comment]
@kazachka , to fix wercker CI , please rebase on latest master code

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java [hunk 3]
```diff
@@ -102,10 +97,10 @@ public void setMax(int max) {
             TokenTypes.LITERAL_IF,
             TokenTypes.LITERAL_ELSE,
             TokenTypes.LITERAL_SWITCH,
-            TokenTypes.LITERAL_CASE,
-            TokenTypes.LITERAL_TRY,
-            TokenTypes.LITERAL_CATCH,
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java]: @romani I don't know NPath but should try/catch be removed?
I assume `case` was converted to `case_group` as I see it counting cases below.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java (hunk 2 above)
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java]: Your fix was fine, I requested it to be in different PR, as problem is different a bit.
Please provide fix separately in scope of #3797.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java (hunk 3 above)
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java]: loosing tokens is not expected, and need to be explained

Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java]: @romani This is from the last PR.
> 4. It doesn't implement NPath calculation for try-catch-finally statements yet. It is a question if we need such calculation. Pmd doesn't implement this calculation and there is no formula in original paper.

Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java]: Possible calculations:
http://stackoverflow.com/questions/31755570/calculate-npath-complexity-for-try-catch-finally?rq=1 (unanswered)
https://github.com/pdepend/pdepend/blob/master/src/main/php/PDepend/Metrics/Analyzer/NPathComplexityAnalyzer.php#L469-L509 (php way)

Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java]: lets, not mix problems. if try-catch does not work, I now try to remember some private discussions with original contributor about this, I will try to find what is left. 
But still, lets do try-catch separately, separate issue.

Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java]: all changes and discussion about TRY forms should be done in new issue #3814 .
Current PR should not change anything for try-catch in functionality.

**Comment by romani** [comment]
@kazachka , please rebase on latest master and remove any changes to try-catch tokens from changes it will addressed in #3814 .

**Comment by kazachka** [comment]
done

**Comment by romani** [comment]
@kazachka ,
please keep CI green.

**Comment by kazachka** [comment]
done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java [hunk 4]
```diff
@@ -51,18 +51,28 @@
...
 
-    /** Stack of values - all but the current value. */
-    private final Deque<BigInteger> valueStack = new ArrayDeque<>();
+    /**
+     * Stack of NP values for ranges.
+     */
+    private final Deque<BigInteger> rangeValues = new ArrayDeque<>();
+
+    /** Stack of NP values for expressions. */
+    private final Deque<Integer> expressionValues = new ArrayDeque<>();
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java]: Why Integer ? (Strange that another stack of BigInteger)

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java [hunk 5]
```diff
@@ -155,56 +196,249 @@ public void leaveToken(DetailAST ast) {
...
+     * Calculates number of logical operators, including inline ternary operatior, for a token.
+     * @param ast inspected token.
+     * @return number of logical operators.
+     */
+    private static int countLogicalOperators(DetailAST ast) {
+        int number = 0;
+        for (DetailAST child = ast.getFirstChild(); child != null;
+                child = child.getNextSibling()) {
+            final int type = child.getType();
+            if (type == TokenTypes.LOR || type == TokenTypes.LAND) {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java]: Why other logical operators are missed ? "^"  ?

Comment (kazachka) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java]: "^" is bitwise operation

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java (hunk 4 above)
Comment (kazachka) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java]: Because NPath value for one expression is small number, unlike NPath value for whole file that can be huge.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java (hunk 5 above)
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java]: https://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.22.2

"Boolean Logical Operators &, ^, and |" vs "Conditional Operator" so what actually you mean in this method ?

**Comment by romani** [comment]
@rnveach , please review, I have only minor issue left.
I did not question design aspect of Check as I want to wait till other issues are fixed for this Check to see whole picture of required logic.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java (hunk 4 above)
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java]: ok, npath above 2,147,483,647 (Integer limit) is smth really strange. And does not mean any useful metric. BigInterger will work always but it is performance hit.
Ok, lets keep it as is for now.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java (hunk 5 above)
Comment (kazachka) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java]: LOR and LAND are || and && conditional operators, but in TokenTypes javadoc they are called as logical.

Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java]: Please create separate issue to fix javadoc, we do not need bad terms that misleading us. Please use correct terms.

Comment (kazachka) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java]: renamed to countConditionalOperators, javadoc fixed

Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java]: All information that was missed and helped you to write correct code should be placed  to javadoc, next contributor might trap in the same problem.
Link to jdk should be in javadoc

Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java]: In  xdoc there is phrase 
> It takes into account the nesting of conditional statements and multi-part boolean expressions (e.g., A && B, C || D, etc.). 

Looks like etc is "?:" only. Let's update xdoc too to be exact

**Comment by kazachka** [comment]
Diff report
https://kazachka.github.io/npath-complexity-diff-report/

**Comment by romani** [comment]
@kazachka , please use latest version of diff report generator that is based on groovy, as it show more details in report.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java (hunk 5 above)
Comment (kazachka) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java]: updated

**Comment by romani** [comment]
from NPATH algorithm description (NPathComplexity-article-p188-nejmeh.pdf document):

> NP(expression) = number of && and || operators in the expression.

so all `&` `|` `^` should be skipped from consideration.

@kazachka , please do not forget to provide new diff report.

**Comment by kazachka** [comment]
https://kazachka.github.io/npath-complexity-diff-report/android-launcher/
https://kazachka.github.io/npath-complexity-diff-report/apache-jsecurity/
https://kazachka.github.io/npath-complexity-diff-report/apache-struts/
https://kazachka.github.io/npath-complexity-diff-report/guava/

**Comment by kazachka** [comment]
https://kazachka.github.io/npath-complexity-diff-report/

**Comment by rnveach** [comment]
@kazachka CI needs to be green.

**Comment by romani** [comment]
@kazachka , from report is not visible how you calculate NPATH for https://kazachka.github.io/npath-complexity-diff-report/apache-jsecurity/index.html#A1

please regenerate report and place max to 1. So all structures will be violated that is expected but due to diff nature of result report you will have the same result but in report we will see all NPATH calculations.
Please expand calculation of apache-jsecurity/index.html#A1 manually to recheck that NPATH is calculated correctly.

**Comment by kazachka** [comment]
https://kazachka.github.io/npath-complexity-diff-report
new diff with max=1

**Comment by romani** [comment]
@kazachka , please rebase, there are some conflicts .

**Comment by kazachka** [comment]
? : operator calculates as 2 + count of child conditional operators
for example:
condition1 ? condition2 : condition3;//2
(condition1 || condition2) ? true : false;//3
(condition1 || condition2) ? (condition2 || condition4): (condition3 || condition4 || condition5);//6

**Comment by kazachka** [comment]
https://kazachka.github.io/npath-complexity-diff-report/apache-jsecurity/index.html#A1
now check calculate like this:
npath = 1 + 1 + 1 + 1 + 1 + 1 = 6
public Object nullSafeGet(ResultSet rs, String[] names, Object owner)
        throws HibernateException, SQLException {
    Object identifier = type.get(rs, names[0]);
    if (identifier == null || rs.wasNull()) { //1 + 1 + 1(for return)
        return null;
    }
    try {//1
        return valueOfMethod.invoke(enumClass, identifier);//1
    } catch (Exception exception) {//1
        String msg = "Exception while invoking valueOfMethod [" + valueOfMethod.getName() +
                "] of Enum class [" + enumClass.getName() + "] with argument of type [" +
                identifier.getClass().getName() + "], value=[" + identifier + "]";
        throw new HibernateException(msg, exception);
    }
}

**Comment by romani** [comment]
> rebase

please rebase one more time, you still have problems with already fixed issues.

> now check calculate like this: npath = 1 + 1 + 1 + 1 + 1 + 1 = 6

that is not correct.
formula demands multiplication of all function/block elements, not a sum.
This should be fixed.

> now ? : operator calculates as 2 + count of child conditional operators

that is a simple case. In example I pointed there is case of expression with in-lined "?:" . 
Unfortunately it is not covered by NPATH pdf specification. We could postpone it to another issue.

> try {//1

that is ok , but not correct.
minal paths through tr-catch block is 2 (without exception and with exception). But this again another issue.

**Comment by kazachka** [comment]
error in calculation
https://github.com/checkstyle/checkstyle/pull/3785#issuecomment-281172607

npath = (1 + 1 + 1) * (1 + 1) = 6
```
public Object nullSafeGet(ResultSet rs, String[] names, Object owner)
        throws HibernateException, SQLException {
    Object identifier = type.get(rs, names[0]);
    if (identifier == null || rs.wasNull()) { //1 + 1 + 1(for return)
        return null;
    }
    try {//1 (for try-return)
        return valueOfMethod.invoke(enumClass, identifier);
    } catch (Exception exception) {//1
        String msg = "Exception while invoking valueOfMethod [" + valueOfMethod.getName() +
                "] of Enum class [" + enumClass.getName() + "] with argument of type [" +
                identifier.getClass().getName() + "], value=[" + identifier + "]";
        throw new HibernateException(msg, exception);
    }
}
```

**Comment by kazachka** [comment]
All the CI's are green.

**Comment by romani** [comment]
@kazachka ,

1) I did  comment `Please update all Inputs with such trace. To make it clear how we are calculation is done.`
please address it in your inputs.

2)
please regenerate diff report , I need to see value for "createBitmapThumbnail" it should be as I described.

**Comment by romani** [comment]
@kazachka , you have commit  but no comments , is it ready for review ?

**Comment by kazachka** [comment]
That commit is to done first point.
I update check (locally), but still have some errors in calculation with LITERAL_ELSE

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/InputNPathComplexity.java [hunk 6]
```diff
@@ -0,0 +1,162 @@
+package com.puppycrawl.tools.checkstyle.checks.metrics;
+// Advise: for lack of ambiguity try to make all factors prime numbers
+public class InputNPathComplexity {
+    //NP = 1 + 4 = 5
+    void testIfWithExpression() {
+        //1 (default way + 4 (for conditionals)
+        if (true && true || (true || true)) { }
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/InputNPathComplexity.java]: this Check is very complicated so lets follow formulas that we defined in our 
http://checkstyle.sourceforge.net/config_metrics.html#NPathComplexity

> NP(if-range) + NP(expr) + 1
P(i=1:i=N)NP(Statement(i))

so comment over IF should be `(if-range=1) + 1 + (expr=4) = 6` . So 6 is npath value for IF only.
NPath value for method is 6 so comment should be `NP = 6`
Please keep this notation to clearly define to what part of code certain value is corresponding.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/InputNPathComplexity.java [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    }
+    
+    //NP = 1 + 4 = 5
+    void testIfElseWithExpression() {
+        //1 + 4 (for conditionals)
+        if (true && true || (true || true)) { }
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/InputNPathComplexity.java]: >  	NP(if-range) + NP(expr) + NP(else-range)

(if-range=1) + (else-range=1) +  (expr=4) = 6

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/InputNPathComplexity.java [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        else { }
+    }
+    
+    //NP = 1 + 3 = 4
+    void testSimpleSwitch() {
+        int a = 0;
+        //1 (default) + 3 (cases)
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/InputNPathComplexity.java]: > S(i=1:i=n)NP(case-range(i)) + NP(default-range) + NP(expr)

( (case-range(1)=1) + (case-range(2)=1) + (case-range(3)=1) ) + (default-range=0) + (expr = 0) = 3

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/InputNPathComplexity.java [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    
+    //NP = 1 + 4 = 5
+    void testIfElseWithExpression() {
+        //1 + 4 (for conditionals)
+        if (true && true || (true || true)) { }
+        else { }
+    }
+    
+    //NP = 1 + 3 = 4
+    void testSimpleSwitch() {
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/InputNPathComplexity.java]: > P(i=1:i=N)NP(Statement(i))

NP= (Statement(1)=1) * (Statement(2)=3) = 3

**Comment by kazachka** [comment]
1. Done
2. https://kazachka.github.io/npath-complexity-diff-report/

**Comment by kazachka** [comment]
NPath for andoroid-launcher Utilities#createBitmapThumbnail is now 30.
https://kazachka.github.io/npath-complexity-diff-report/android-launcher/index.html#A201
```
static Bitmap createBitmapThumbnail(Bitmap bitmap, Context context) { // 2 * 1 * 15 = 30
 if (sIconWidth == -1) {  // 1 + 0 + 1 = 2
     foo(); // 1
 }
 
 foo();  // 1
 
 if (width > 0 && height > 0) { // (13 from below) + 1 + 1 = 15 ^
     if (width < bitmapWidth || height < bitmapHeight) { // (1 * 3 * 3 * 1 * 1) + 1 + 3 = 13 ^ 
         foo(); //1
     
         if (bitmapWidth > bitmapHeight) { // 1 + 0 + (2 from below) = 3 
             foo();
         } else if (bitmapHeight > bitmapWidth) { //1 + 0 + 1 = 2 ^
             foo();
         }
     
         final Bitmap.Config c = (width == sIconWidth && height == sIconHeight) ? // 1 + 0 + 0 + 2 = 3 ^
                 bitmap.getConfig() : Bitmap.Config.ARGB_8888;
         foo();  //1
         return thumb;  // 1
     } else if (bitmapWidth < width || bitmapHeight < height) { // 1 + 1 + 1 = 3 ^
         foo();
         return thumb;
     }
 }
}
```

**Comment by romani** [comment]
@kazachka , please remove extra log file from commit.
Please reply "done" to each my point for your previous commit.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/InputNPathComplexity.java [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+    void testIfElseWithExpression() {
+        // (if-range=1) + (else-range=1) + (expr=3) = 5
+        if (true && true || (true || true)) { }
+        else { }
+    }
+    
+    //NP = 4
+    int testSimpleSwitch() {
+        int a = 0;
+        // NP = (case-range(1)=1) + (case-range(1)=1) + (case-range(1)=1)
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/InputNPathComplexity.java]: 1) please change "case-range(1)" to "case-range[1]" - lets use index syntax to avoid ambiguity. I updated xdoc file.
2) please fix index values in other cases
3) please apply 1)-2) to all Input files. If structure has strict names for parts then array indexes are not required. Array indexes are only for cases[i] and statement[i]

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/InputNPathComplexity.java [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+        for (int i = 0; i < 10; i++);
+        // NP(for-statement(2)) = (for-range=1) + (expr(1)=0) + (expr(2)=1) + (expr(3)=0) = 2
+        for (int i = 0; i < 10 && true; i++);
+        // NP(for-statement(3)) = (for-range=1) + (expr(1)=2) + (expr(2)=0) + (expr(3)=0) = 3
+        for (int i = true ? 0 : 0; i < 10; i++);
+        // NP(for-statement(4)) = (for-range=1) + (expr(1)=0) + (expr(2)=1+2) + (expr(3)=0) = 4
+        for (int i = 0; true ? i < 10 : true || true; i++);
+    }
+    
+    // NP = (do-statement(1)=5) * (do-statement(2)=2) = 12
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/InputNPathComplexity.java]: NP = (do-statement[1]=5) * (do-statement[2]=2) = 12

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/InputNPathComplexity.java [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+    void testDoWhileCyclesComplex() {
+        int a = 0;
+        // NP(do-statement(1)) = (do-range=1) + (expr=1) = 2
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/metrics/InputNPathComplexity.java]: `NP(do-statement) = (do-range=1) + (expr=1) = 2` , do-statement without index
